Use System.nanoTime() when building Duration objects

petroav opened the pull request:
Favor System.nanoTime() to System.currentTimeMillis() because
System.currentTimeMillis() is not monotonic.

petroav commented:
Fixes #3599.  I grepped the codebase for all uses of `System.currentTimeMillis()`.

petroav commented on presto-main/src/main/java/com/facebook/presto/server/QueryExecutionResource.java:
```diff
@@ -102,7 +102,7 @@ public Response getTaskInfo(@PathParam("queryId") String queryId)
                 }
 
-                long last = System.currentTimeMillis();
+                long last = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
```
I changed this because even though it's not directly used in a `Duration` object it is used to measure a duration on line 113. I traced the second term in that subtraction to originate from `System.nanoTime()` and thought that the first term should also derive from `System.nanoTime()`.

cberner commented:
Thanks for taking this on! Can you also check for instances that indirectly derive from a non-monotonic clock? For example, line 149 in `QueryStateMachine` is not safe because it subtracts two values derived from `DateTime.now()`: `elapsedTime = new Duration(endTime.get().getMillis() - createTime.getMillis(), MILLISECONDS);`

electrum commented:
Please remove the leading "*" and extra field "Testing" from your commit message.

petroav commented:
@electrum yep, sure thing!

Sorry I didn't give this attention over the past week. I was busy with other stuff but now I've freed up and should provide an update soon.

petroav commented:
Rebased on master and addressed @electrum's commit message comments.

I did a more thorough investigation this time, manually going through all the ways in which you can build a `Duration` object: `nanosSince()`, `succinctNanos()`, `succinctDuration()` and `new Duration()`. For each usage I ensured that the origin of the arguments could be traced back to a monotonically increasing clock. I didn't find any blatantly wrong usage other than the changes included in this PR. I paid particular attention to the following classes: `QueryStateMachine`, `QueryStats`, `StageStateMachine`, `StageStats`, `TaskContext`, `TaskStats`, `PipelineContext`, `PipelineStats`, `DriverContext`, `DriverStats` and `OperatorContext`. If there are other classes that you know might be susceptible to this problem, I'd be happy to go and check.

The only other dubious calculation of elapsed time I found was in `OperatorContext.recordGeneratedInput()`. Maybe we can add a check to set `readNanos` to zero if we ever get passed a negative value? I examined the current usages of the method and it seems that all calculations are done safely.

martint commented:
I made that minor change and merge it. Thanks!
